### PR TITLE
Remove more initial PHP state

### PR DESCRIPTION
**Pull request by noahtallen**
## Description
This PR looks at removing more initial PHP state for the site editor. Since things can be loaded more async now, I shored up many selectors and code paths to make sure they can handle null/undefined values.

This may also set us up more for loading different dynamic pages on the fly.

I will add comments where some of my concerns are.

## How has this been tested?
Locally in edit site.

## Types of changes
Code quality

## Checklist:
- [ ] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR.

### CODE REVIEW

File: lib/edit-site-page.php [hunk 1]
```diff
@@ -145,17 +145,8 @@ function gutenberg_edit_site_init( $hook ) {
...
-	$settings['editSiteInitialState']['templateType'] = 'wp_template';
-	$settings['editSiteInitialState']['showOnFront']  = get_option( 'show_on_front' );
-	$settings['editSiteInitialState']['page']         = array(
-		'path'    => '/',
-		'context' => 'page' === $settings['editSiteInitialState']['showOnFront'] ? array(
-			'postType' => 'page',
-			'postId'   => get_option( 'page_on_front' ),
-		) : array(),
-	);
+	$settings['showOnFront'] = get_option( 'show_on_front' );
```
Comment (noahtallen) [inline lib/edit-site-page.php:148]: I could not determine a way to fetch these options from the client without adding extra overrides. It would be helpful to have a way to fetch arbitrary options from the client. I'm not sure why these are not included in the default settings.

File: packages/edit-site/src/components/save-button/index.js [hunk 2]
```diff
@@ -22,7 +22,6 @@ export default function SaveButton( { openEntitiesSavedStates } ) {
 			isSaving: some( dirtyEntityRecords, ( record ) =>
 				isSavingEntityRecord( record.kind, record.name, record.key )
 			),
-			templateType: select( 'core/edit-site' ).getTemplateType(),
```
Comment (noahtallen) [inline packages/edit-site/src/components/save-button/index.js:25]: This was not used in the file.

File: packages/edit-site/src/components/template-switcher/index.js [hunk 3]
```diff
@@ -69,20 +68,7 @@ export default function TemplateSwitcher( {
 		setThemePreviewVisible( () => false );
 	};
 
-	const registry = useRegistry();
-	const [ homeId, setHomeId ] = useState();
-
-	useEffect( () => {
-		findTemplate(
-			'/',
-			registry.__experimentalResolveSelect( 'core' ).getEntityRecords
```
Comment (noahtallen) [inline packages/edit-site/src/components/template-switcher/index.js:78]: We store the homepage template ID in the state now.

File: packages/edit-site/src/store/actions.js [hunk 4]
```diff
@@ -107,19 +107,39 @@ export function setTemplatePart( templatePartId ) {
...
 	};
 }
+
+/**
+ * Sets up the initial template and page for edit site, and fetches other
+ * information it needs.
+ *
+ * @param {Object} initialPage The initial page to load in the site editor.
+ */
+export function* setupState( initialPage ) {
```
Comment (noahtallen) [inline packages/edit-site/src/store/actions.js:134]: This doesn't feel like a great place to put this logic. Where else could it go?

File: packages/edit-site/src/store/index.js [hunk 5]
```diff
@@ -23,8 +23,20 @@ export default function registerEditSiteStore( initialState ) {
 		initialState,
 	} );
 
+	const { showOnFront, pageOnFront } = initialState.home;
+	const initialPage = {
```
Comment (noahtallen) [inline packages/edit-site/src/store/index.js:27]: Ultimately, it would be nice to be able to set this value externally from Gutenberg. It would also be nice to just pass in `initialRoute: '/'` (or even a page ID) and then edit site resolves everything else which is needed.

File: lib/edit-site-page.php (hunk 1 above)
Comment (noahtallen) [reply lib/edit-site-page.php:148]: I think we can just register these as settings now.

File: packages/edit-site/src/components/editor/index.js [hunk 6]
```diff
@@ -125,25 +129,31 @@ function Editor() {
...
 		} ),
-		[ page.context ]
+		[ page?.context ]
 	);
+
+	let entityId;
+	if ( templateType ) {
+		entityId = templateType === 'wp_template' ? templateId : templatePartId;
+	}
+
```
Comment (ockham) [inline packages/edit-site/src/components/editor/index.js:156]: Maybe move this before R85 so we can use `entityId` on [R89-R91](https://github.com/WordPress/gutenberg/pull/23775/files#diff-641afe1a0152b0e5136c8b46edb84b33R89-R91)?

File: packages/edit-site/src/store/actions.js (hunk 4 above)
Comment (ockham) [reply packages/edit-site/src/store/actions.js:134]: TBH I'm generally not a big fan of actions that set initial state. I prefer reducers with a reasonable initial state; and, if we need to fetch information for a more meaningful de-facto initial state, I'd prefer an action that also makes sense to set state at a later time.

File: packages/edit-site/src/store/reducer.js [hunk 7]
```diff
@@ -169,24 +158,32 @@ export function page( state = {}, action ) {
...
  *
  * @param {Object} state Current state.
+ * @param {Object} action Dispatched action.
  *
  * @return {Object} Updated state.
  */
-export function showOnFront( state ) {
+export function home( state = {}, action ) {
+	switch ( action.type ) {
+		case 'SET_HOME_TEMPLATE':
```
Comment (ockham) [inline packages/edit-site/src/store/reducer.js:170]: In a similar vein as https://github.com/WordPress/gutenberg/pull/23775/files#r451417455 -- a `home` reducer (for setting the home page template), and corresponding `SET_HOME_TEMPLATE` is quite specific.

While it certainly is information that's quite relevant for the site editor, it raises the question why we aren't storing the same information for other pages (or more generally, categories etc). I think we made a decision to not store a lot of template/template parts related information in a dedicated state subtree, but use `getEntitiesRecord` and friends :thinking:

File: packages/edit-site/src/store/selectors.js [hunk 8]
```diff
@@ -88,18 +88,18 @@ export const getSettings = createSelector(
...
+	return state.home.templateId;
 }
 
 /**
  * Returns the current template ID.
  *
  * @param {Object} state Global application state.
  *
- * @return {number} Template ID.
+ * @return {number?} Template ID.
```
Comment (ockham) [inline packages/edit-site/src/store/selectors.js:102]: Wondering which underlying change is allowing for this to be `undefined`. Shouldn't there always be a 'current' template?

File: packages/edit-site/src/store/index.js (hunk 5 above)
Comment (ockham) [reply packages/edit-site/src/store/index.js:27]: > Ultimately, it would be nice to be able to set this value externally from Gutenberg.

Not sure I'm following, can you elaborate?

> It would also be nice to just pass in `initialRoute: '/'` (or even a page ID) and then edit site resolves everything else which is needed.

I think that's kind of the point of `findTemplate` -- pass a route, and get the associated template.

> (or even a page ID)

I think a route is preferable, since it can be used for non-page entities (categories etc)?

File: packages/edit-site/src/store/actions.js (hunk 4 above)
Comment (ockham) [reply packages/edit-site/src/store/actions.js:134]: I can maybe rephrase to -- what does `setupState` give us that `setPage` doesn't?

**Review by ockham — CHANGES_REQUESTED** [review]
Left a few notes. I _think_ that we might be able to drop a few more of the ancillary functions that have been introduced here as replacements for the previous initial state and tackle the problem at a higher level (hopefully):

I think the mechanism that we have in place to get the template for a given page should be relatively stable; we might not even need the extra logic involving `show_on_front`/`page_on_front`, but maybe we can basically just pass in the `/` route and rely on what is already there for template resolution (basically `findTemplate` and the related backend 'endpoint' doing all the heavy lifting).

File: packages/edit-site/src/store/actions.js (hunk 4 above)
Comment (noahtallen) [reply packages/edit-site/src/store/actions.js:134]: it fetches and stores the homepage template, which is needed to display which "template" is the homepage. this can be different from the initial page if we ever want to change which page is loaded at start

I agree with you in general, and wanted to block initializing the editor on async data load (which is why this is not part of the initial state).

It sounds like we just want to run a couple of extra actions when initializing the editor, rather than calling a setup action?

File: packages/edit-site/src/store/index.js (hunk 5 above)
Comment (noahtallen) [reply packages/edit-site/src/store/index.js:27]: > Not sure I'm following, can you elaborate?

1. deep linking into the site editor to access a specific page
2. from a different integration of the site editor, telling it the initial page/route to load

basically, just allowing you to modify the default page loaded by the editor

> I think that's kind of the point of findTemplate

`findTemplate` doesn't return the required page/post ID which would be needed to load the post content.

> I think a route is preferable

I tend to agree :) I was thinking of a situation where we know the post ID / page ID and want to tell the site editor to load that post/page

File: packages/edit-site/src/store/reducer.js (hunk 7 above)
Comment (noahtallen) [reply packages/edit-site/src/store/reducer.js:170]: homepage data isn't stored anywhere in gutenberg yet, and we need a way to know what the homepage is while looking at other pages on the site. (e.g. to show the "home" icon" beside the homepage in the page switcher)

File: packages/edit-site/src/store/selectors.js (hunk 8 above)
Comment (noahtallen) [reply packages/edit-site/src/store/selectors.js:102]: This was already the case, but the template ID can be undefined since we load the template async at the beginning. Additionally, if anything errors out, we may end up without a "current template" (e.g. that "empty document" situation). So I wanted to change the type so that we handle this scenario safely without crashing

**Comment by noahtallen** [comment]
> we might not even need the extra logic involving show_on_front/page_on_front

@ockham we do need to know what the post_id is (page_on_front) :) Though I'm not sure, does findTemplate give us that?

File: packages/edit-site/src/components/editor/index.js (hunk 6 above)
Comment (noahtallen) [reply packages/edit-site/src/components/editor/index.js:156]: Hm, maybe we can add an entityId to the select callback and then it could be available there

File: packages/edit-site/src/components/editor/index.js [hunk 9]
```diff
@@ -125,25 +129,31 @@ function Editor() {
 	// and Query Pagination blocks.
 	const blockContext = useMemo(
 		() => ( {
-			...page.context,
-			query: page.context.query || { categoryIds: [] },
+			...page?.context,
```
Comment (jeyip) [inline packages/edit-site/src/components/editor/index.js:132]: [TIL about the null propagation operator](https://ponyfoo.com/articles/null-propagation-operator) 

Very cool!

File: packages/edit-site/src/store/reducer.js [hunk 10]
```diff
@@ -169,24 +158,32 @@ export function page( state = {}, action ) {
...
- * Reducer returning the site's `show_on_front` setting.
+ * Reducer for information about the site's homepage.
  *
  * @param {Object} state Current state.
+ * @param {Object} action Dispatched action.
  *
  * @return {Object} Updated state.
  */
-export function showOnFront( state ) {
+export function home( state = {}, action ) {
```
Comment (jeyip) [inline packages/edit-site/src/store/reducer.js:168]: It seems like reducer `state` would always be defined here. Is there an instance where we need the default param?

Comment (noahtallen) [reply packages/edit-site/src/store/reducer.js:168]: I think you're right, but I like to add it in case we need to "home" from the initial state passed into registerStore

File: packages/edit-site/src/components/editor/index.js (hunk 6 above)
Comment (jeyip) [reply packages/edit-site/src/components/editor/index.js:156]: That makes sense to me.

Since `entityId` is derived from values that are available in `useSelect` anyway, we could consolidate the logic in there.

File: packages/edit-site/src/components/page-switcher/index.js [hunk 11]
```diff
@@ -111,8 +111,8 @@ export default function PageSwitcher( {
 			toggleProps={ {
 				children:
 					[ ...pages, ...categories, ...posts ].find(
-						( choice ) => choice.value === activePage.path
-					)?.label || activePage.path,
+						( choice ) => choice.value === activePage?.path
```
Comment (jeyip) [inline packages/edit-site/src/components/page-switcher/index.js:114]: **Non-blocking**:

What do we think about consolidating instances of `activePage?.path`? We could define `const activePagePath = activePage?.path` and replace R114, R115, etc. with references to the variable?

**Review by jeyip — COMMENTED** [review]
I see that we have unit tests `edit-site` [selectors](https://github.com/WordPress/gutenberg/blob/c3d21df412b7fd554dde1cd6478c30390ac5554c/packages/edit-site/src/store/test/selectors.js), [reducer](https://github.com/WordPress/gutenberg/blob/c3d21df412b7fd554dde1cd6478c30390ac5554c/packages/edit-site/src/store/test/reducer.js), and [actions](https://github.com/WordPress/gutenberg/blob/c3d21df412b7fd554dde1cd6478c30390ac5554c/packages/edit-site/src/store/test/actions.js). Should we update those in this PR?

File: packages/edit-site/src/components/page-switcher/index.js (hunk 11 above)
Comment (noahtallen) [reply packages/edit-site/src/components/page-switcher/index.js:114]: sounds good!

**Comment by noahtallen** [comment]
yeah, will do! (wanted to make sure the code was in a good place before modifying them)

File: packages/edit-site/src/store/actions.js (hunk 4 above)
Comment (ockham) [reply packages/edit-site/src/store/actions.js:134]: > it fetches and stores the homepage template, which is needed to display which "template" is the homepage. this can be different from the initial page if we ever want to change which page is loaded at start

I'll play the YAGNI card here -- let's not prepare for that case before it arrives :slightly_smiling_face: 

> I agree with you in general, and wanted to block initializing the editor on async data load (which is why this is not part of the initial state).
> 
> It sounds like we just want to run a couple of extra actions when initializing the editor, rather than calling a setup action?

I think I'd overall prefer that architecturally, yeah :thinking:

**Comment by noahtallen** [comment]
@ockham, Let me know what you think of the new approach. I added a `showHomepage` action which computes and sets the homepage. I think this is better than a general "setup state" action since it's more modular. (We now load show on front/page on front over the API)

File: packages/edit-site/src/store/actions.js [hunk 12]
```diff
@@ -105,21 +105,62 @@ export function setTemplatePart( templatePartId ) {
...
+ * @param {string}  page.slug    The page slug.
+ * @param {string}  page.path    The page path.
+ * @param {Object}  page.context The page context.
+ * @param {number?} templateId   The template ID for the page if we already know it.
  *
  * @return {Object} Action object.
  */
-export function* setPage( page ) {
-	const templateId = yield findTemplate( page.path );
+export function* setPage( page, templateId ) {
```
Comment (ockham) [inline packages/edit-site/src/store/actions.js:132]: From a separation of concerns POV, I'm still not too happy we're passing `templateId` to `setPage`.
Since it's obtained via `findTemplate` both here, and in `showHomepage` -- is there any chance this is being cached at some level anyway?

File: packages/edit-site/src/components/header/index.js [hunk 13]
```diff
@@ -49,16 +49,22 @@ export default function Header( {
 			getTemplatePartId,
 			getTemplateType,
 			getPage,
-			getShowOnFront,
 		} = select( 'core/edit-site' );
+
+		// eslint-disable-next-line camelcase
```
Comment (ockham) [inline packages/edit-site/src/components/header/index.js:54]: Maybe there's a convention that we're really supposed to return camel case from the endpoint here?

File: packages/edit-site/src/store/actions.js (hunk 12 above)
Comment (ockham) [reply packages/edit-site/src/store/actions.js:132]: We could alternatively absorb some of the extra logic from `showHomepage` into this function, and run the `show_on_front`/`page_on_front` stuff only `if ( page === '/' )` :thinking:

**Comment by ockham** [comment]
A few unit tests are failing and need updating I think :sweat_smile:

File: packages/edit-site/src/components/header/index.js (hunk 13 above)
Comment (noahtallen) [reply packages/edit-site/src/components/header/index.js:54]: The existing data stored here doesn't seem to be in camel case:

Screen Shot 2020-07-13 at 11 22 43 AM

I'm not sure what API convention is here, though. At the very least, I think it needs to use underscores when passed into the API, but I don't know when it needs to be converted

File: packages/edit-site/src/store/actions.js (hunk 12 above)
Comment (noahtallen) [reply packages/edit-site/src/store/actions.js:132]: The request does happen twice without this code, but we should definitely try to cache it properly! I'm not sure where to start to make that happen, though.

Screen Shot 2020-07-13 at 11 39 45 AM

> We could alternatively absorb some of the extra logic from showHomepage into this function

Yeah, that could work too. That means that we aren't always relying solely on the passed `page` object, though.

Alternatively, we could add a `setRoute` function which:
- computes required page context (like query/page ID)
- computes required template ID
- sets the page & template

and then we do `setRoute( '/' );`

But that doesn't solve the problem of passing the templateId to setPage.

We could always let it fetch twice, but it'd be nice to avoid the performance hit

File: packages/edit-site/src/components/header/index.js (hunk 13 above)
Comment (noahtallen) [reply packages/edit-site/src/components/header/index.js]: I've renamed the variable inline. I think that's good enough

File: packages/edit-site/src/store/actions.js (hunk 12 above)
Comment (noahtallen) [reply packages/edit-site/src/store/actions.js]: Ok, I've added a naive cache layer to `findTemplate`. I think this solves the problem well enough for now (though we might need to think about the consequences a bit more).

Ideally, (per @epiqueras) we would add `path` to the getEntityRecords function. Then it could be passed to the actual WordPress query. The problem here is that it would be fairly complicated to create a `get_template_from_path` function server-side because template resolution is tied so closely with page render.

**Comment by noahtallen** [comment]
Alright, @ockham, tests are good now. One challenging aspect of the cache is that there could be edge cases around `removeTemplate` and `addTemplate`. So it may be worth just... not optimizing for this case yet. But that means we'll fetch the homepage twice at the beginning, which adds an extra second or two of delay. Any ideas?

File: packages/edit-site/src/store/actions.js (hunk 12 above)
Comment (ockham) [reply packages/edit-site/src/store/actions.js]: > Ok, I've added a naive cache layer to `findTemplate`. I think this solves the problem well enough for now (though we might need to think about the consequences a bit more).

Thanks! I'm unfortunately still a bit concerned about it living at the module level :sweat_smile: 
We've repeatedly seen issues in other projects that import `@wordpress/` dependencies with some of their singleton-style data management (specifically when different versions of one `@wordpress/` package -- most notably `@wordpress/data` I think -- ended up in our npm dependency tree without getting de-duplicated).

So I'd rather steer clear of module-level caching and the like. I noticed that we're [creating a registry control for the `findTemplate` util](https://github.com/WordPress/gutenberg/blob/1ba985d852dbb2ca8d91befe2b30a552fe251794/packages/edit-site/src/store/controls.js#L26-L31) in [`controls.js`](https://github.com/WordPress/gutenberg/blob/1ba985d852dbb2ca8d91befe2b30a552fe251794/packages/edit-site/src/store/controls.js). I'm not too familiar with controls' (and registries') internals, but this was one of the reasons I was hoping we'd be some caching for free (I do seem to remember that caching for network requests is added in a similar way in `wp/core` or `core-data`.) More importantly, since we're essentially wrapping the util in a control (that is tied to a registry), I was thinking that we _should_ at least be caching at registry level -- I think there can multiple registries, so I think that each should have its individual cache.

My apologies, I truly don't mean to stall this PR. Maybe we should remove the caching for now and add it separately in a subsequent PR? I feel that I personally don't know enough about controls and registries, and that there might be a 'right' way to do it that I have yet to learn about. I'm a bit uncomfortable with the module level caching and would lean towards removing it.

> Ideally, (per @epiqueras) we would add `path` to the getEntityRecords function. Then it could be passed to the actual WordPress query. The problem here is that it would be fairly complicated to create a `get_template_from_path` function server-side because template resolution is tied so closely with page render.

Yep! That'd be the holy grail. :+1:

Comment (noahtallen) [reply packages/edit-site/src/store/actions.js]: > I was hoping we'd be some caching for free

I think we do get caching for free as long as we can use `getEntityRecords` all the time for `findTemplate`. The current problem is that `getEntityRecord` cannot query for a template by path, so we are stuck doing our own ad-hoc request for it :(

> I truly don't mean to stall this PR. Maybe we should remove the caching for now and add it separately in a subsequent PR?

This sounds perfectly acceptable to me! Given that we want to spend more time thinking about it, it deserves its own PR.

**Comment by epiqueras** [comment]
You don't need to call `findTemplate` in `showHomepage`. Just have `setPage` return the template ID.

**Comment by noahtallen** [comment]
> Just have setPage return the template ID.

Nice, I think this is a decent way to solve the problem for now. Done in 1e47384

File: packages/edit-site/src/store/actions.js (hunk 12 above)
Comment (noahtallen) [reply packages/edit-site/src/store/actions.js]: How do you feel about just returning the resolved ID from `setPage`? it works well and doesn't really compromise on separation of concerns

**Comment by noahtallen** [comment]
Will need to do some more work on the e2e test tomorrow, but it should be good to go other than that. Things test well in the editor

**Comment by ockham** [comment]
The failing e2e test seems similar to one that was failing with #22893, where I was also removing [`$settings['editSiteInitialState']['templateType'] = 'wp_template';`](https://github.com/WordPress/gutenberg/pull/23775/files#diff-5022ec1d311d9e1661b75192ce0385aaL150). Once I added that line back, the test passed (and I could remove a number of other e2e test workarounds that had added `waitForSelector`).

Maybe some UI elements (the template part dropdown?) that we query by selector in the e2e tests become available while others don't yet (the actual list of template parts), and the e2e test tries to look for them too early. A simple fix could be to set the `templateType` reducer's initial state to `wp_template` -- that's basically the same thing we were doing via PHP before.

**Comment by ockham** [comment]
> A simple fix could be to set the `templateType` reducer's initial state to `wp_template`

This seems to fix the failing e2e test in my local testing. I'll modify unit tests accordingly, and push a commit. I might also be able to roll back e2e test changes :crossed_fingers:

File: packages/edit-site/src/store/actions.js (hunk 12 above)
Comment (ockham) [reply packages/edit-site/src/store/actions.js]: WFM!

**Comment by ockham** [comment]
Post editor snapshot tests are failing, which is unrelated to this PR (and consistent with #23848). I'll rebase.

**Comment by ockham** [comment]
> Post editor snapshot tests are failing, which is unrelated to this PR (and consistent with #23848). I'll rebase.

Well, that didn't work. Looks like this might be broken in `master`, since a lot of current PRs have the same issue.

**Comment by ockham** [comment]
> > Post editor snapshot tests are failing, which is unrelated to this PR (and consistent with #23848). I'll rebase.
> 
> Well, that didn't work. Looks like this might be broken in `master`, since a lot of current PRs have the same issue.

Tracked it down and filed an issue: #23996

**Comment by noahtallen** [comment]
Thanks for looking into that, @ockham! I think I was okay making the change to use `waitForSelector` because I think data load of the initial template happens more async now, so my guess was that it had to wait longer before things showed up. Adding that to the initial state makes sense to me though!

**Comment by ockham** [comment]
> > > Post editor snapshot tests are failing, which is unrelated to this PR (and consistent with #23848). I'll rebase.
> > 
> > 
> > Well, that didn't work. Looks like this might be broken in `master`, since a lot of current PRs have the same issue.
> 
> Tracked it down and filed an issue: #23996

This has been fixed by #24004. I'll rebase again :crossed_fingers:

**Comment by noahtallen** [comment]
Nice, tests are looking good.

**Review by ockham — APPROVED** [review]

File: lib/edit-site-page.php [hunk 14]
```diff
@@ -225,3 +213,33 @@ function gutenberg_edit_site_init( $hook ) {
...
+		array(
+			'show_in_rest' => array(
+				'name' => 'page_on_front',
+			),
+			'type'         => 'number',
+			'description'  => __( 'The ID of the page that should be displayed on the front page', 'gutenberg' ),
+		)
+	);
+}
+add_action( 'rest_api_init', 'register_site_editor_homepage_settings', 10 );
```
Comment (youknowriad) [inline lib/edit-site-page.php:245]: I believe this is adding data to some REST API endpoint. So pinging @TimothyBJacobs for awareness.

Comment (TimothyBJacobs) [reply lib/edit-site-page.php:245]: Thanks for the ping! Doing this on `rest_api_init` should work, but it might be better to do it on `init` so it is always a registered setting. We can also safely drop the `name` parameter for `show_in_rest`. It'll default to the option name.

Comment (noahtallen) [reply lib/edit-site-page.php:245]: Nice, I will definitely make those changes.

Comment (noahtallen) [reply lib/edit-site-page.php:245]: https://github.com/WordPress/gutenberg/pull/24390
